Math/NumberTheory/Primes/Factorisation/QuadraticSieve.hs

```diff
@@ -0,0 +1,143 @@
+{-# LANGUAGE CPP #-}
+
+module Math.NumberTheory.Primes.Factorisation.QuadraticSieve
+  ( quadraticSieve
+  , findPairs
+  ) where
+
+#if __GLASGOW_HASKELL__ < 803
+import Data.Semigroup
+#endif
+import qualified Data.Vector as V
+import qualified Data.Vector.Mutable as MV
+import qualified Data.IntMap as I
+import qualified Data.IntSet as S
+import qualified Math.NumberTheory.Primes.IntSet as PS
+import Control.Monad
+import Control.Monad.ST
+import Data.Maybe
+import Data.Bifunctor
+import Math.NumberTheory.Roots
+import Math.NumberTheory.Primes
+import Math.NumberTheory.Moduli.Sqrt
+import Math.NumberTheory.Utils.FromIntegral
+
+data SignedPrimeIntSet = SignedPrimeIntSet
+  { sign :: !Bool
+  , primeSet :: !PS.PrimeIntSet
+  } deriving (Show)
+
+data BoolOrPrime = Bool Bool | PrimeInt (Prime Int)
+
+insert :: Prime Int -> SignedPrimeIntSet -> SignedPrimeIntSet
+insert prime (SignedPrimeIntSet s ps) = SignedPrimeIntSet s (prime `PS.insert` ps)
+
+nonZero :: SignedPrimeIntSet -> Maybe BoolOrPrime
+nonZero (SignedPrimeIntSet s ps) = case PS.minView ps of
+  Just (prime, _)  -> Just (PrimeInt prime)
+  Nothing          -> if s then Just (Bool True) else Nothing
+
+member :: BoolOrPrime -> SignedPrimeIntSet -> Bool
+member value (SignedPrimeIntSet s ps) = case value of
+  Bool b     -> s == b
+  PrimeInt p -> p `PS.member` ps
+
+xor :: SignedPrimeIntSet -> SignedPrimeIntSet -> SignedPrimeIntSet
+xor (SignedPrimeIntSet s1 ps1) (SignedPrimeIntSet s2 ps2) = SignedPrimeIntSet (s1 /= s2) ((ps1 PS.\\ PS.unPrimeIntSet ps2) <> (ps2 PS.\\ PS.unPrimeIntSet ps1))
```

Comment:
Just to elaborate a little on what I have in mind: convert to both sets to lists, e.g. `PS.toList` and then traverse them booth at the same time kind of like we do during merge sort and then reconstruct the result as a set. There's `Data.IntSet.fromDistinctAscList` function which should do this with reasonably small overhead.

Reply:
I tried today this approach, but it appeared to be slower. Seems, bitmasks in `IntSet` are quite efficient (even while `xor` is 3x slower than it could be, if it was a built-in operation).